File: midirectestimate2.m
function MI_est = ...
    midirectestimate2(yData,nTrials,nBoots,binCounts,subSamps)
%MIDIRECTESTIMATE computes a direct estimate of mutual information through
%iterative naive estimates.
% 
% Inputs
%   > yData:
%   > xData:
%   > nTrials:
%   > nBoots:
%   > binCountArray: 
%   > subsampleArray:
% Outputs
%   > MI_est: nTrials x nBinSizes x nSubsamplesets x nBoots array of naive
%   estimates. To be extrapolated to infinitte data and 0 bin size.
%
% Based on Eq. 7 and Eq. 24 in Tkacik et al, 2015.
 
[nBinsX,nE,nG] = size(yData);
x = 1/nBinsX:1/nBinsX:1;

assert(nBinsX == numel(x));

% nX = 100; % For pSmad data with such small sample size.

m = round(subSamps*nE);
nSubsamplesets = numel(m);

replacement = false;

embryoIdcs = [1:1:nE];

nBinSizes = numel(binCounts);

for iTrial = 1:nTrials
    % Initialize this trial.
    MI_est(iTrial).binSizes = binCounts;
    MI_est(iTrial).subSamps = subSamps;
    MI_est(iTrial).nBoots = nBoots;
    
    MI_est(iTrial).naiveEst_means = zeros(nBinSizes,nSubsamplesets);
    MI_est(iTrial).naiveEst_stds = zeros(nBinSizes,nSubsamplesets);
    if nTrials > 1
%         disp(['Beginning trial ',num2str(iTrial),'/',num2str(nTrials)])
%         tic
    end
    iBinSizeIdx = 0;
    for iBinSize = 1:nBinSizes
        if nTrials == 1
%             disp(['Beginning bin ',num2str(iBinSize),'/',num2str(nBinSizes)])
%             tic
        end
        nBinsG = binCounts(iBinSize);
        
        nBinsX = nBinsG;
        pxBin = 1/nBinsX; % Assume uniform distribution
        
        iBinSizeIdx = iBinSizeIdx + 1;
        for iSubsampleIdx = 1:nSubsamplesets
            % Initialize distributions for subsample size.
            mSubsamps = m(iSubsampleIdx);
            pg = zeros(nBinsG,nBinsG); % signal distribution         
            pgxJ = zeros(nBinsX, nBinsG,nBinsG); % joint distribution
            MIEq24 = zeros(nBoots,1);
            for iBoot = 1:nBoots
                k = m(iSubsampleIdx);
                subIdx = randsample(embryoIdcs,k,replacement);
                sub1 = yData(:,subIdx,1);
                sub2 = yData(:,subIdx,2);

                sub1 = reshape(sub1,numel(sub1),1);
                sub2 = reshape(sub2,numel(sub2),1);

                xMat = repmat(x',1,mSubsamps);
                xColumn = reshape(xMat,numel(xMat),1);
                
                %% Distributions
                pgxJ = histcountsn([xColumn,sub1,sub2],[nBinsX,nBinsG,nBinsG], ...
                    'Normalization', 'probability');
                pgxJ = pgxJ + eps;
                pgxJ_perm = permute(pgxJ,[2,3,1]);

                % P(g1,g2): Joint PDF of {g1,g2}
                pg = squeeze(sum(pgxJ,1));

                % P(x): PDF of x (uniform)
                px = repmat(pxBin,nBinsX,1);                                  
                
                % P({g1,g2} | x): Joint PDF estimate of {g1,g2} conditional on x 
                pg_given_x = pgxJ ./ px;
                pg_given_x_perm = permute(pg_given_x,[2,3,1]);
                
                %% Eq. 7
                % Inner summation/intergraion term
                logValue = pg_given_x_perm .* log2(pg_given_x_perm ./ pg);
                
                MIEq7_temp = sum(px .* sum(logValue,3),'all');
                
                MIEq7(iBoot) = MIEq7_temp;
                 
                %% Eq.24
                MIEq24_temp = ...
                    sum( pgxJ_perm .* log2(pgxJ_perm ./ (px .* pg)), 'all' );  
                
                MIEq24(iBoot) = MIEq24_temp;
                
                %% Validation
                assert(abs(MIEq7_temp - MIEq24_temp) < 1e-10)
                             
            end
                        
            MI_est(iTrial).naiveEst_means(iBinSize,iSubsampleIdx) = ...
                mean(MIEq24);
            MI_est(iTrial).naiveEst_stds(iBinSize,iSubsampleIdx) = ...
                std(MIEq24,0);
            
        end
        
        if nTrials == 1
%             toc
%             disp([num2str(round(100*iBinSize/nBinSizes,2)),'%'])
        end
    end
    
    if nTrials > 1
%         toc
%         disp([num2str(round(100*iTrial/nTrials,2)),'%'])
    end
end

%save(['direct-mi-estimate'],'MI_est','nBoots','binCounts','subSamps','m','nE')

File: midirectestimate2.py
import numpy as np
import random
import math

def midirectestimate2(yData,nTrials,nBoots,binCounts,subSamps):
    
    nBinsX = yData.shape[0]
    nE = yData.shape[1]
    nG = yData.shape[2]

    x=np.arange(1/nBinsX, 1, 1/nBinsX) # start: 1/nBinsX, stop: 1, step: 1/nBinsX

    assert (nBinsX == np.size(x))

    # nX = 100; % For pSmad data with such small sample size.

    m = round(subSamps*nE)
    nSubsamplesets = np.size(m)

    replacement = False

    embryoIdcs = [np.arange(1, nE, 1)]

    nBinSizes = np.size(binCounts)
    
    MI_est = {}

    for iTrial in range(0,nTrials):
        
        MI_est[iTrial] = {'binSizes':binCounts,'subSamps':subSamps,'nBoots':nBoots,'naiveEst_means': (np.zeros(nBinSizes,nSubsamplesets)),'naiveEst_stds':(np.zeros(nBinSizes,nSubsamplesets))}
        
    
        if nTrials > 1:
            print("Beginning trial " + str(iTrial) + "/" + str(nBinSizes))
            #tic
            
            iBinSizeIdx =0
            
            for iBinSize in range(0,nBinSizes):
                if nTrials == 1:
                    print("Beginning bin " + str(iBinSize) + "/" + str(nBinSizes))
                    #tic
            
            nBinsG = binCounts[iBinSize]
        
            nBinsX = nBinsG
            pxBin = 1/nBinsX
        
            iBinSizeIdx = iBinSizeIdx + 1
            
            for iSubsampleIdx in range(0, nSubsamplesets):
                mSubsamps = m[iSubsampleIdx]
                pg = np.zeros(nBinsG, nBinsG)
                pgxJ = np.zeros(nBinsX, nBinsG,nBinsG)
                MIEq24 = np.zeros(nBoots, 1)
                for iBoot in range(0, nBoots):
                    subIdx = random.choices(embryoIdcs, weights=None, k=mSubsamps)
                    sub1 = yData[:,subIdx, 1]
                    sub2 = yData[:,subIdx, 2]
                    
                    sub1 = np.reshape(sub1, np.size(sub1))
                    sub2 = np.reshape(sub2, np.size(sub2))
                    
                    xMat = np.tile(x.conj().transpose(), (1, mSubsamps))
                    xColumn = np.reshape(xMat, np.size(xMat))
                    
                    #Distributions
                    #pgxJ = histcountsn([xColumn, sub],[nBinsX,nBinsG], ...
                    #'Normalization','probability'); %joint probability distribution of G and X
                    #Use new histcounts function
                    
                    
                    pgxJ =pgxJ + eps
                    pgxJ_perm = np.transpose(pgxJ(2,3,1))
                    
                    pg = np.sum(pgxJ,1).squeeze()
                    px = np.tile(pxBin, (nBinsX, 1))
                    
                    pg_given_x = pgxJ / px
                    pg_given_x_perm = np.transpose(pg_given_x,(2,3,1))
                    
                    #Eq. 7
                    logValue = pg_given_x_perm  * math.log2((pg_given_x_perm /pg))
                    
                    MIEq7_temp = np.sum(px * np.sum(logValue,3))
                
                    MIEq7[iBoot] = MIEq7_temp
                    
                    #Eq. 24
                    MIEq24_temp = np.sum(pgxJ_perm * math.log2(pgxJ_perm / (px * pg)))
                
                    MIEq24[iBoot] = MIEq24_temp
                    
                    #Validation
                    assert(abs(MIEq7_temp - MIEq24_temp) < 1e-10)
                
            
            if nTrials ==1:
                print(str(round(100*iBinSize/nBinSizes,2)))
                
        if nTrials>1:
            print(str(100*iTrial/nTrials,2))
            
    #save file
                    
                    
        
    
    return MI_est